### ISSUE TITLE
Use JSONSerialization instead of Yams for JSON

### ISSUE THREAD

**Issue by marcelofabri** [SwiftGen/SwiftGen#698]
Right now, JSON parsing is done via `Yams.load_all` since JSON is a subset of Yaml. 

That's clever, but comes with a perfomance cost. For my project, I'm using SwiftGen to parse Lottie JSON animation files into enums and that's taking almost ~3s per (incremental build).

My initial thought was to add caching support (#695). But I think we can avoid that (for now?) and do something way simpler: using `JSONSerialization` instead. In my small test, this made `swiftgen` to take ~750ms instead.

This is what I did - I can clean this and submit a PR, just need some guidance:

```swift
// MARK: JSON File Parser

public enum JSON {
  public enum ParserError: Error, CustomStringConvertible {
    case invalidFile(path: Path, reason: String)

    public var description: String {
      switch self {
      case .invalidFile(let path, let reason):
        return "Unable to parse file at \(path). \(reason)"
      }
    }
  }

  public class Parser: SwiftGenKit.Parser {
    var files: [File] = []
    public var warningHandler: Parser.MessageHandler?

    public required init(options: [String: Any] = [:], warningHandler: Parser.MessageHandler? = nil) {
      self.warningHandler = warningHandler
    }

    public class var defaultFilter: String {
      return ".*\\.(?i:json)$"
    }

    public func parse(path: Path, relativeTo parent: Path) throws {
      files.append(try File(path: path, relativeTo: parent))
    }
  }

  // is it ok to duplicate this struct? I did it to avoid breaking any public APIs, but I guess I could extract it to another type and have a typealias here. Thoughts?
  struct File {
    let path: Path
    let name: String
    let documents: [Any]

    init(path: Path, relativeTo parent: Path? = nil) throws {
      guard let data: Data = try? path.read() else {
        throw ParserError.invalidFile(path: path, reason: "Unable to read file")
      }

      self.path = parent.flatMap { path.relative(to: $0) } ?? path
      self.name = path.lastComponentWithoutExtension

      do {
        let item = try JSONSerialization.jsonObject(with: data)
        self.documents = [item] // why does the Yaml version take an array? is this the right thing for JSON?
      } catch let error {
          throw ParserError.invalidFile(path: path, reason: error.localizedDescription)
      }
    }
  }
}


// this is pretty much the same as `Yaml.Parser`. What is the right abstraction here? Are we ok with duplicating it?
extension JSON.Parser {
  public func stencilContext() -> [String: Any] {
    let files = self.files
      .sorted { lhs, rhs in lhs.name < rhs.name }
      .map(map(file:))

    return [
      "files": files
    ]
  }

  private func map(file: JSON.File) -> [String: Any] {
    return [
      "name": file.name,
      "path": file.path.string,
      "documents": file.documents.map(map(document:))
    ]
  }

  private func map(document: Any) -> [String: Any] {
    return [
      "data": document,
      "metadata": Metadata.generate(for: document)
    ]
  }
}

```

**Comment by AliSoftware** [comment]
That seems like a great suggestion, and indeed would be easier at least as first step than implementing cache.

To answer your questions:
 - I think you took the right approach from what I read in your sample code 👍 
 - YAML files can contain multiple "documents" (separated by `---`) inside them – see [example in test fixture here](https://github.com/SwiftGen/SwiftGen/blob/master/Tests/Fixtures/Resources/YAML/good/documents.yaml). That's why its `File` structure has an array for `documents: [Any]`.
 - This means that I think it's ok to duplicate the `struct File` for the JSON parser, and modify it to replace `documents: [Any]` by `content: Any` instead, since unlike YAML, JSON files can't contain multiple documents
 - That would also allow us to simplify the templates for JSON, as currently in those we're iterating over `documents` – or are just using `documents.first`, I don't remember which – as we know that JSON only have one document. With that change we'd now be able to use `content` instead of `documents.first` which would make the JSON template more understandable
 - Similarly for the `extension JSON.Parser` I think it's ok to keep it separate from the YAML one because with that change from `documents: [Any]` to `content: Any`, that extension will have to be adapted for the JSON case too anyway

---

Going forward I suggest we do that in 2 steps:

1. Do a first PR which would target 6.2.1 or 6.3 and contain, basically, exactly what you already implemented in this PR's description
    - keep the API intact (i.e. keep `documents: [Any]` in the implementation and have no changes in the templates
    - but use `JSONSerialization` instead of `Yams`
2. Then do a second PR on top of that, targeting the next major version 7.0, which would:
   - change `documents: [Any]` to `content: Any`
   - update the contexts documentation (currently the [Yaml Context doc](https://github.com/SwiftGen/SwiftGen/blob/master/Documentation/SwiftGenKit%20Contexts/Json.md) is a symlink of the [JSON Context doc](https://github.com/SwiftGen/SwiftGen/blob/master/Documentation/SwiftGenKit%20Contexts/Yaml.md) because they're exactly the same, so with that change we'd replace the symlink with an amended copy of the doc instead)
    - update the built-in JSON templates accordingly

Would that work for you?

**Comment by marcelofabri** [comment]
> Would that work for you?

Thanks for the detailed explanation! I've opened #702 against `master` which should be covering **1**. 

Do we have a 7.0.0 branch? I was assuming `develop` was going to be used for 6.3 based on other issues.

**Comment by AliSoftware** [comment]
We follow git-flow type of branching strategy, so:
 - master is only containing the merge commits of tagged, released versions
 - develop is for active development. Any new PR is supposed to start from `develop`
 - We don't create `release/*` branches until we are ready to release (they typically only contain potentially cleaning the CHANGELOG entries and updating the podspecs)

So I think the best approach here for your case would be to create the PR for (2) on top of #702 and specify explicitly that it should only be merged in 7.0 – we have a GitHub label `breaking` for that, as well as GitHub milestones already created for 7.0. And switch its base branch to `develop` once #702 is merged.

Sure that would mean that we might need to rebase / update that future PR once we're ready to tackle 7.0, as intermediate PRs would likely have been merged in develop in-between, but even if we already had a release branch that would still be needed anyway ;)
